src/main/scala/overflowdb/codegen/CodeGen.scala

```diff
@@ -427,38 +413,105 @@ def writeConstants(outputDir: JFile): JFile = {
+            }
+          }
+        }.mkString("\n")
 
         s""" {
         |  val properties = new JHashMap[String, AnyRef]
-        |  $putKeysImpl
+        |$putKeysImpl
+        |$putRefsImpl
         |  properties
         |}""".stripMargin
       }
 
-      val containedNodesAsMembers =
-        nodeType.containedNodesList.map { containedNode =>
-          val containedNodeType = containedNode.nodeTypeClassName
-          val cardinality = Cardinality.fromName(containedNode.cardinality)
-          val completeType = cardinality match {
-            case Cardinality.ZeroOrOne => s"Option[$containedNodeType]"
-            case Cardinality.One       => containedNodeType
-            case Cardinality.List      => s"List[$containedNodeType]"
-          }
-          val traversalEnding = cardinality match {
-            case Cardinality.ZeroOrOne => s".headOption"
-            case Cardinality.One       => s".head"
-            case Cardinality.List      => s".toList"
+      val fromNew = {
+        val initKeysImpl = keys
+          .map { key: Property =>
+            val memberName = camelCase(key.name)
+            Cardinality.fromName(key.cardinality) match {
+              case Cardinality.One =>
+                s"""   this._$memberName = other.$memberName""".stripMargin
+              case Cardinality.ZeroOrOne =>
+                s"""   this._$memberName = if(other.$memberName != null) other.$memberName else None""".stripMargin
+              case Cardinality.List =>
+                s"""   this._$memberName = if(other.$memberName != null) other.$memberName else Nil""".stripMargin
+            }
+            s"""   this._$memberName = other.$memberName""".stripMargin
           }
+          .mkString("\n")
+        val initRefsImpl = {
+          nodeType.containedNodesList.map { containedNode =>
+            val memberName = containedNode.localName
+            val containedNodeType = containedNode.nodeTypeClassName
 
-          s"""/** link to 'contained' node of type $containedNodeType */
-             |def ${containedNode.localName}: $completeType =
-             |  edges(Direction.OUT, "${DefaultEdgeTypes.ContainsNode}").asScala.toList
-             |    .filter(_.valueOption(EdgeKeys.LOCAL_NAME).map(_  == "${containedNode.localName}").getOrElse(false))
-             |    .sortBy(_.valueOption(EdgeKeys.INDEX))
-             |    .map(_.inVertex.asInstanceOf[$containedNodeType])
-             |    $traversalEnding
-             |""".stripMargin
+            Cardinality.fromName(containedNode.cardinality) match {
+              case Cardinality.One =>
+                s"""  this._$memberName = other.$memberName match {
+                   |    case null => null
+                   |    case newNode: NewNode => mapping(newNode).asInstanceOf[$containedNodeType]
+                   |    case oldNode: StoredNode => oldNode.asInstanceOf[$containedNodeType]
+                   |    case _ => throw new MatchError("unreachable")
```

Comment:
This case is indeed only to shut up the compiler, since it cannot prove that no other cases exist. Hence the very descriptive error message and choice of exception ;)

Reply:
:+1: